### PR TITLE
Remove shared nsg rule immediately when no destinations prefixes are left

### PR DESCRIPTION
**Pull request by MartinForReal**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Remove shared nsg rule immediately when no destinations left
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Fixed: Remove shared nsg rule immediately when no destinations left
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure_loadbalancer.go [hunk 1]
```diff
@@ -2658,28 +2658,35 @@ func (az *Cloud) reconcileSecurityRules(sg network.SecurityGroup, service *v1.Se
...
+							newDestinations := append(existingPrefixes[:addressIndex], existingPrefixes[addressIndex+1:]...)
+							sharedRule.DestinationAddressPrefixes = &newDestinations
+							updatedRules[sharedIndex] = sharedRule
+						}
+						dirtySg = true
+					}
+				}
+
+				if shouldDeleteNSGRule {
 					klog.V(4).Infof("Didn't find DestinationAddressPrefixes in shared rule for service %s", service.Name)
```
Comment (jwtty) [inline pkg/provider/azure_loadbalancer.go:2685]: We should change this log

File: pkg/provider/azure_loadbalancer_test.go [hunk 2]
```diff
@@ -3785,6 +3785,52 @@ func TestReconcileSecurityGroup(t *testing.T) {
 				},
 			},
 		},
+		{
```
Comment (jwtty) [inline pkg/provider/azure_loadbalancer_test.go:3788]: Do we have a test case where the sg rule should be deleted completely?

Comment (MartinForReal) [reply pkg/provider/azure_loadbalancer_test.go:3788]: Please refer to https://github.com/kubernetes-sigs/cloud-provider-azure/blob/3f6d95e591a14322599495c6ffd30f68f6aa7d1c/tests/e2e/network/network_security_group.go#L143-L191

File: pkg/provider/azure_loadbalancer.go (hunk 1 above)
Comment (MartinForReal) [reply pkg/provider/azure_loadbalancer.go]: updated. Thanks!

**Review by feiskyer — COMMENTED** [review]
/lgtm

**Pull request merged**